### ISSUE TITLE
Calling Variant() dereferences freed and invalid object pointers, leading to potential use-after-frees.

### ISSUE THREAD

**Issue by id01** [godotengine/godot-cpp#1387]
### Godot version

4.2.1.stable

### godot-cpp version

4.2.1.stable

### System information

Arch Linux, Windows 10

### Issue description

Hey y'all, hope you're ready for this one because it's a banger!

#### The Problem

I started getting some weird crashes on Windows exports; no stack trace, no nothing. Opened it in WinDbg and dumped the stack trace. Happened in a callback to the Godot binary in the `free()` function of my GDExtension. (relevant portion of WinDbg traceback; the rest is trash about embree due to lack of debug symbols in the Godot binary used):
```
libshipgame_simulation_windows_template_debug_x86_64!ZN28ShipBuildBlockInfoCollection4freeEPvS0_+0x74
```

If you open this portion up in IDA, the return address points to the function here:
```
call    _ZN5godot7VariantC2EPKNS_6ObjectE
```

Here, a `godot::Variant()` is called before passing it into a `UtilityFunctions::is_instance_valid()` in the compiled code.

Here's the relevant portion of the GDCLASS macro which defines `free()` for all classes. After executing the destructor, it calls free_static on cls, which according to the disassembler, seems to have an implicit cast converting cls into a Variant.
```
	static void free(void *data, GDExtensionClassInstancePtr ptr) {                                                                                                                    \
		if (ptr) {                                                                                                                                                                     \
			m_class *cls = reinterpret_cast<m_class *>(ptr);                                                                                                                           \
			cls->~m_class();                                                                                                                                                           \
			::godot::Memory::free_static(cls);                                                                                                                                         \
		}                                                                                                                                                                              \
	}
```

On Linux, this had no issue, but on Windows, I just happened to have a class destructor that called `queue_free()` on some elements that also happened to be children of the class. It seems that on Windows, the freeing happened during the call, which meant that this issue was causing the free() function to be called twice, and the failsafe with the `is_instance_valid` is dereferencing the freed memory, causing a crash. As to why queue_free seems to be freeing the memory here, I'll have to look into further.

Looking at the godot-cpp code further, it seems like we're passing this over to the main Godot binary, which matches my observations, but the Variant() constructor in the main Godot binary dereferences the Variant object without actually checking if it is valid.

Variant constructor in godot-cpp:
```
Variant::Variant(const Object *v) {
	if (v) {
		from_type_constructor[OBJECT](_native_ptr(), const_cast<GodotObject **>(&v->_owner));
	} else {
		GodotObject *nullobject = nullptr;
		from_type_constructor[OBJECT](_native_ptr(), &nullobject);
	}
}
```

Variant setter for object in main Godot binary:
```
	_FORCE_INLINE_ static void object_assign(Variant *v, const Variant *o) {
		object_assign(v, o->_get_obj().obj);
	}
```

Due to this, Variant() was dereferencing an invalid (freed) pointer which led to a memory access violation.

#### Impact

First of all, the double-free on Windows is an issue that would be patched by this, but we also have issues with when users of godot-cpp want to handle freed objects. In GDScript, a freed object can be manipulated, but will cause errors when referenced, and say `[previously freed]` when printed. In godot-cpp, we probably don't want casting a freed node to a Variant to crash the application, particularly because `UtilityFunctions::is_instance_valid()` casts the freed node to a Variant.

#### Possible Solutions

Possible solutions for this problem depend on how you want to handle invalid pointers. If it's okay to convert invalid pointers to null pointers in godot-cpp, a workaround could simply be to add a `UtilityFunctions::is_instance_valid`-esque call to the verification (note that is_instance_valid takes a Variant pointer), but this would mean that `[previously freed]` instances become null instances, which may not be intuitive for application logic.

The other solution is to add the check in the main Godot binary, creating an invalid instance with the `[previously freed]` that we recognize from GDScript. I'm not particularly sure about the internals of previously freed instances, but this would be more consistent with GDScript.

Thank you for listening to my TED talk

### Steps to reproduce

You can create an invalid pointer using reinterpret_cast. Freed pointers are just invalid pointers which may or may not point to valid memory.

```
class Test : public Node {
	GDCLASS(Test, Node);
private:
	Variant ptr;

protected:
	static void _bind_methods() {
		ClassDB::bind_method(D_METHOD("test"), &Test::test);
	}

public:
	Test() {}
	~Test() {}

	void test() { ptr = Variant(reinterpret_cast<Node*>(1)); UtilityFunctions::print(ptr); }
};
```

Expected result:
`test()` should print `[invalid object]`, `[previously freed]`, or `<null>`

Actual result:
Program crashes

This may cause issues in scenarios like the below:

```
Node* ptr = reinterpret_cast<Node*>(1); // Invalid ptr, from free() or otherwise
if (UtilityFunctions::is_instance_valid(ptr)) { // Crash!! Implicit cast to Variant()
   ...do things...
}
```

### Minimal reproduction project

Compile the code in Steps to Reproduce and add it to a GDExtension. Call Test.test(). The program crashes.

**Comment by id01** [comment]
Update:
I also found what's causing the double free on Windows.

Here's some compiled code from the node destructor that appears after the code I wrote.
![image](https://github.com/godotengine/godot-cpp/assets/19232106/4314c4d4-7fb3-496e-9c89-a68237d22e11)

I see the same compiled code in the free() function for the same class, after already calling the destructor for that class.
![image](https://github.com/godotengine/godot-cpp/assets/19232106/b9d9f99b-8ef9-4ee4-9b11-419463292197)

It seems like somehow, on the Mingw compile, the code to free the class is being put both in the free() function and the destructor function, which causes a double free.

It seems like this is a composite between two issues - the is_instance_valid issue and the Windows double free issue.
